### DISCUSSION TITLE
A time to consider how to improve the efficiency of building the menu nodes

### DISCUSSION THREAD

**Discussion by marksweb** [django-cms/django-cms#7805] [General]
It's been raised on [discord](https://discord.com/channels/800813886689247262/1204047551570120755) that the `_build_nodes` process from the menu system can cause problems for projects with many pages/languages.

To save people looking to discord for the details;

> We have migrated to 4.1 and have a large number of pages (18k) across 17 languages. I have noticed an issue when starting up the server. 
> The admin pages are available instantly. But there is a huge lag (hour +) when viewing a page in a language for the first time. Once the page is eventually rendered all subsequent pages that havent been loaded before are fine (no lag).
> If I was then to manually adjust the url of a pages that has been rendered there is a similar length lag. Again once a language has been loaded for the first time all other pages in that language are fine. 
> I set the logging to the maximum level and can see what is causing this in the console. When loading a page/language combo for the first time this query is fired:
```sql
 SELECT 
    cms_pagecontent.id, 
    cms_pagecontent.language, 
    cms_pagecontent.title, 
    cms_pagecontent.page_title, 
    cms_pagecontent.menu_title, 
    cms_pagecontent.meta_description, 
    cms_pagecontent.redirect, 
    cms_pagecontent.page_id, 
    cms_pagecontent.creation_date, 
    cms_pagecontent.created_by, 
    cms_pagecontent.changed_by, 
    cms_pagecontent.changed_date, 
    cms_pagecontent.in_navigation, 
    cms_pagecontent.soft_root, 
    cms_pagecontent.template, 
    cms_pagecontent.limit_visibility_in_menu, 
    cms_pagecontent.xframe_options, 
    cms_page.id, 
    cms_page.created_by, 
    cms_page.changed_by, 
    cms_page.creation_date, 
    cms_page.changed_date, 
    cms_page.reverse_id, 
    cms_page.navigation_extenders, 
    cms_page.login_required, 
    cms_page.is_home, 
    cms_page.application_urls, 
    cms_page.application_namespace, 
    cms_page.languages, 
    cms_page.is_page_type, 
    cms_page.node_id, 
    cms_treenode.id, 
    cms_treenode.path, 
    cms_treenode.depth, 
    cms_treenode.numchild, 
    cms_treenode.parent_id, 
    cms_treenode.site_id 
FROM 
    cms_pagecontent 
    INNER JOIN cms_page ON (cms_pagecontent.page_id = cms_page.id) 
    INNER JOIN djangocms_versioning_version ON (cms_pagecontent.id = djangocms_versioning_version.object_id AND (djangocms_versioning_version.content_type_id = 10)) 
    INNER JOIN cms_treenode ON (cms_page.node_id = cms_treenode.id) 
WHERE 
    (cms_pagecontent.language = 'en' AND cms_pagecontent.page_id IN (SELECT U0.id FROM cms_page U0 INNER JOIN cms_treenode U1 ON (U0.node_id = U1.id) WHERE U1.site_id = 1) AND djangocms_versioning_version.state IN ('published')) ORDER BY cms_treenode.path ASC, djangocms_versioning_version.state ASC;
```
> In our case this brings back 18k rows. This query is then run for each row returned:
```
(0.047) SELECT "cms_pageurl"."id", "cms_pageurl"."slug", "cms_pageurl"."path", "cms_pageurl"."language", "cms_pageurl"."page_id", "cms_pageurl"."managed" FROM "cms_pageurl" WHERE ("cms_pageurl"."page_id" = 4810 AND "cms_pageurl"."language" IN ('en', 'ar', 'de', 'es', 'fr', 'he', 'hi', 'it', 'ja', 'ko', 'nl', 'pl', 'pt', 'ru', 'tr', 'zh-hans', 'zh-hant')); args=(4810, 'en', 'ar', 'de', 'es', 'fr', 'he', 'hi', 'it', 'ja', 'ko', 'nl', 'pl', 'pt', 'ru', 'tr', 'zh-hans', 'zh-hant');
```
> Only after all 18k queries have run does the page then render for the first time. As mentioned - once a language has loaded once it is fine. I guess becuase it is not running this mamouth query each time. 
> Is this loading every possible page for each language a configurable option that can be switched off because the volume of pages will only increase and therefore the lag to view a page after starting the server will only continue to expand and get longer.

The code causing this is [here](https://github.com/django-cms/django-cms/blob/develop-4/menus/menu_pool.py#L143) and it know's that it's slow/problematic as the doc string starts...

```python
    def _build_nodes(self):
        """
        This is slow. Caching must be used.
        One menu is built per language and per site.
```

This code has existed as it is since the beginning of the project, coming from [2010](https://github.com/django-cms/django-cms/commit/5c9aa1f3a988b828b6afe300a27097bff9adfefc#diff-9291cc832c240e72dab361e6d38f387c9637501c5f56e85e5315d771df672341R1)

### AI suggestion

With such old code we've no experts in this area, so I've taken some input from Jetbrains AI;

> The _build_nodes function is composed of several steps, each with potential for optimization. Here are some ideas:
> 1. **Caching Improvements:** You already are using cache to store the `final_nodes` result, which is an excellent strategy. Consider using a more fine-grained caching strategy. For example, you could cache individual `menu.get_nodes(self.request)` results, as these calls might also be expensive.
> 2. **Parallelize get_nodes Calls:** If the get_nodes function calls are independent of each other and can be run concurrently, consider using multithreading to run them in parallel. This would greatly improve performance when dealing with a large number of menus.
```
from concurrent.futures import ThreadPoolExecutor

with ThreadPoolExecutor() as executor:
    future_to_menu = {executor.submit(menu.get_nodes, self.request): menu_class_name for menu_class_name in self.menus}
    for future in concurrent.futures.as_completed(future_to_menu):
        menu_class_name = future_to_menu[future]
        try:
            nodes = future.result()
        except NoReverseMatch:
    # rest of your code...
```
> 3. **Optimize `_build_nodes_inner_for_one_menu` Call:** As this function is called inside the for-loop for self.menus, its performance directly impacts the speed of `_build_nodes`. If there are ways to make `_build_nodes_inner_for_one_menu` faster, the entire function will be faster as well.
>  4. **Avoid Unnecessary Operations:** The `CacheKey.objects.create()` operation happens every time the function is called, even if the cache was hit and returned early. Consider moving it up under the condition `if cached_nodes and self.is_cached:` as an `else` clause. This avoids hitting the database needlessly when the cache is hit.

> Before implementing these improvements, make sure to profile your code to identify where the bottlenecks really are. Just because a function is slow, doesn't mean the problem is in that function itself; it may be that it's calling another, super-slow function, or the slowness could be due to I/O, network latency, a slow database query, etc. Use a tool like `cProfile` for profiling Python code.
> Also, always test performance improvements to ensure they're actually having the desired impact. It's common to think something will make code faster, but actually it makes it slower (or has no impact). Use a tool like `timeit` to measure execution time.
> And remember, the key to speeding up a slow function is to find the slowest part of that function and make it faster. No amount of optimizing the fast parts of your function will make up for a small part of it that's super slow.

**Comment by marksweb** [comment]
Point 4 is an interesting one. And is a part of the system that I've never really liked - many years ago the presence of cache keys in the database caused me problems in a load test due to duplicate keys.

The database is being his for cache related matters when `_build_nodes` does `if cached_nodes and self.is_cached:` - `is_cached` does a database `filter`. Then if that doesn't return a result from the database, we do a `CacheKey.objects.create()`

This is creating keys in the format `key = f"{prefix}menu_nodes_{self.request_language}_{self.site.pk}"` so there's not a lot of data in play, just strings.

It'd be worth looking to see if we can collect all these cache key objects together and bulk create them.

**Reply by benzkji** [reply]
The root of everything IMHO is: A query is executed 18k times, because there exist 18k pages. No additional caching, multithreading, or whatever else optimization can help against that? Not at real expert as well, but maybe joining the pageurl table as well, and afterwards operate on the obviously larger queryset might be an option? Just thinking loud...

**Reply by marksweb** [reply]
@benzkji Yes, I haven't got as far as finding the code that actually gets the pages yet. I imagine there will be more to think about when I get to that point.

**Comment by Aiden-RC** [comment]
Just a comment here, this wasn't an issue for us with the same amount of data ( if not more ) on v3, it's only since v4 that we've had this issue, i know djangocms-versioning has it's own CMSMenu class, theres a potential this issue is caused here?

**Reply by benzkji** [reply]
@fsbraun might tell you more. There were efforts to "standardize" menu code, so versioning, non-versioning, and possibly alternate versioning would use the same code.

**Reply by fsbraun** [reply]
I was just looking at this, and to me, it seems that both the `cms_menus` in the core  has and versioning misses a prefetch for the `PageURL` model. 

@Aiden-RC You could check what happens to performance if you set `DJANGOCMS_VERSIONING_ENABLE_MENU_REGISTRATION` to `False`. 

The quick fix might be looking at the queryset here: 

https://github.com/django-cms/djangocms-versioning/blob/8f5c0aef531dec37d191ff2642e95db46815fa53/djangocms_versioning/cms_menus.py#L105-L112 

It seems to me that there's a select related missing for `page__urls`.

**Reply by fsbraun** [reply]
Just to document the resolution:

As of django CMS 4.1.1, the djangocms-version's `CMSMenu` class is not needed and should not be used:
* It does not fully comply with the documented behavior
* It is far less performant

django CMS 4.1.1's `CMSMenu` class is at the performance level of its counterpart in v3 **and** supports versioned pages (by djangocms-versioning or other third-party package).

Here's the PR that fixed the issue: https://github.com/django-cms/django-cms/pull/7807

**Comment by fsbraun** [comment]
I see two performance improvement levers:
* **Move from lists of `NavigationNode` to a dict of `NavigationNode`:**
  *  Since Python 3.6 dicts preserve order (which obviously is important for navigation nodes on the same level)
  * Template tags would not have to iterate through a potentially long (>>10,000) list of nodes before finding the node they need to operate on
* Upon changes to the page tree (or navigation in general), not the whole tree needs to be recalculated but **only the changed node (plus its descendants) can be updated**

(Plus we also have the cache key issue as mentioned by @marksweb)

**Reply by fsbraun** [reply]
Here's a PR that improves menu creation by a factor of 7 or so: #7956. It lays the ground for updating the cached menus, but does not implement this.

Taken together with this fix in djangocms-versioning (https://github.com/django-cms/djangocms-versioning/pull/417), menu creation on my local dev system with sqlite and 50.000 pages takes less than 2 seconds.

**Comment by fsbraun** [comment]
Solved by #7956
